// ng-lazyload-image/esm5/src/lazyload-image.js
import { of } from 'rxjs';
import { catchError, filter, map, mergeMap, take, tap } from 'rxjs/operators';
export function lazyLoadImage(hookSet, attributes) {
    return function (evntObservable) {
        return evntObservable.pipe(tap(function (data) { return attributes.onStateChange.emit({ reason: 'observer-emit', data: data }); }), filter(function (event) {
            return hookSet.isVisible({
                element: attributes.element,
                event: event,
                offset: attributes.offset,
                scrollContainer: attributes.scrollContainer
            });
        }), take(1), tap(function () { return attributes.onStateChange.emit({ reason: 'start-loading' }); }), mergeMap(function () { return hookSet.loadImage(attributes); }), tap(function () { return attributes.onStateChange.emit({ reason: 'mount-image' }); }), tap(function (imagePath) {
            return hookSet.setLoadedImage({
                element: attributes.element,
                imagePath: imagePath,
                useSrcset: attributes.useSrcset
            });
        }), tap(function () { return attributes.onStateChange.emit({ reason: 'loading-succeeded' }); }), map(function () { return true; }), catchError(function (error) {
            attributes.onStateChange.emit({ reason: 'loading-failed', data: error });
            hookSet.setErrorImage(attributes);
            return of(false);
        }), tap(function () {
            attributes.onStateChange.emit({ reason: 'finally' });
            hookSet.finally(attributes);
        }));
    };
}
